src/codegen.c

```diff
@@ -280,77 +282,107 @@ static int add_jump_set(struct codegen_ctxt *ctxt, __u32 what, int tloc,
 
 #define ARG_WORD(arg, word) ((word == HIGH_WORD) ? ARG_HIGH(arg) : ARG_LOW(arg))
 #define NUM_WORD(num, word) ((word == HIGH_WORD) ? NUM_HIGH(num) : NUM_LOW(num))
-
 #define BPF_LOAD_ARCH \
   BPF_STMT(BPF_LD + BPF_W + BPF_ABS, offsetof(struct seccomp_data, arch))
 #define BPF_LOAD_SYSCALL \
   BPF_STMT(BPF_LD + BPF_W + BPF_ABS, offsetof(struct seccomp_data, nr))
 #define BPF_LOAD_ARG_WORD(arg, high) \
   BPF_STMT(BPF_LD + BPF_W + BPF_ABS, ARG_WORD(arg, high))
 
-static uint32_t value_of(struct expr_tree *expr, int word);
-
 static bool is_const_value(struct expr_tree *expr, int word) {
-  switch (expr->type) {
-    case EXPR_NUMBER:
-      return true;
-    case EXPR_VAR:
-      return false;
-    case EXPR_BIT_AND:
-      if (is_const_value(expr->right, word) &&
-          value_of(expr->right, word) == 0) {
-        return true;
-      }
-      if (is_const_value(expr->left, word) && value_of(expr->left, word) == 0) {
-        return true;
-      }
-      return false;
-    default:
-      ASSERT(0);  // should not happen
-  }
+  return word == HIGH_WORD ? expr->high.is_const : expr->low.is_const;
 }
 
 static uint32_t value_of(struct expr_tree *expr, int word) {
+  return word == HIGH_WORD ? expr->high.value : expr->low.value;
+}
+
+static void cache_constants_by_word(struct expr_tree *expr, int word) {
+  struct cached_value *cached = (word == HIGH_WORD) ? &expr->high : &expr->low;
+  uint32_t clobber_value = 0;
   switch (expr->type) {
     case EXPR_NUMBER:
-      return NUM_WORD(expr->number, word);
+      cached->is_const = true;
+      cached->value = NUM_WORD(expr->number, word);
+      return;
+    case EXPR_VAR:
+      cached->is_const = false;
+      return;
+    case EXPR_BIT_OR:
+      clobber_value = UINT32_MAX;
+      /* fall-through */
     case EXPR_BIT_AND:
+      cache_constants_by_word(expr->right, word);
+      cache_constants_by_word(expr->left, word);
       if (is_const_value(expr->right, word) &&
-          value_of(expr->right, word) == 0) {
-        return 0;
+          value_of(expr->right, word) == clobber_value) {
+        cached->is_const = true;
+        cached->value = clobber_value;
+      } else if (is_const_value(expr->left, word) &&
+          value_of(expr->left, word) == clobber_value) {
+        cached->is_const = true;
+        cached->value = clobber_value;
+      } else {
+        cached->is_const = false;
       }
-      if (is_const_value(expr->left, word) && value_of(expr->left, word) == 0) {
-        return 0;
-      }
-    // fall-through
+      return;
     default:
-      ASSERT(0);  // should not happen
+      cache_constants_by_word(expr->right, word);
```

Comment:
`ASSERT(expr->type >= EXPR_BINARY_MIN && expr->type <= EXPR_BINARY_MAX)` just in case.